zipkintracing/tracing.go

```diff
@@ -0,0 +1,221 @@
+func (t *tracer) TraceFunc(c echo.Context, spanName string, tags map[string]string) func() {
+	span, _ := t.tracer.StartSpanFromContext(c.Request().Context(), spanName)
+	for key, value := range tags {
+		span.Tag(key, value)
+	}
+	finishSpan := func() {
+		span.Finish()
+	}
+
+	return finishSpan
+}
+
+func (t *tracer) GetTraceProxyFunc(tags map[string]string) echo.MiddlewareFunc {
+	return t.GetTraceProxyFuncWithConfig(tags, DefaultTraceProxyFuncConfig)
+}
+
+// GetTraceProxyFunc takes desire tags and returns middleware used to trace proxy calls
+func (t *tracer) GetTraceProxyFuncWithConfig(tags map[string]string, config TraceProxyFuncConfig) echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			if config.Skipper(c) {
+				return next(c)
+			}
+			var parentContext model.SpanContext
+			if span := zipkin.SpanFromContext(c.Request().Context()); span != nil {
+				parentContext = span.Context()
+			}
+			span := t.tracer.StartSpan(fmt.Sprintf("C %s %s", c.Request().Method, "reverse proxy"), zipkin.Parent(parentContext))
+			for key, value := range tags {
+				span.Tag(key, value)
+			}
+			defer span.Finish()
+			ctx := zipkin.NewContext(c.Request().Context(), span)
+			c.Request().WithContext(ctx)
+			b3.InjectHTTP(c.Request())(span.Context())
+			nrw := NewResponseWriter(c.Response().Writer)
+			next(c)
+
+			if nrw.Size() > 0 {
+				zipkin.TagHTTPResponseSize.Set(span, strconv.FormatInt(int64(nrw.Size()), 10))
+			}
+			if nrw.Status() < 200 || nrw.Status() > 299 {
+				statusCode := strconv.FormatInt(int64(nrw.Status()), 10)
+				zipkin.TagHTTPStatusCode.Set(span, statusCode)
+				if nrw.Status() > 399 {
+					zipkin.TagError.Set(span, statusCode)
+				}
+			}
+			return nil
+		}
+	}
+}
+
+func (t *tracer) GetTraceServerFunc(tags map[string]string) echo.MiddlewareFunc {
+	return t.GetTraceServerFuncWithConfig(tags, DefaultTraceServerFuncConfig)
+}
+
+// GetTraceServerFunc takes desire tags and returns middleware used to trace server call
+func (t *tracer) GetTraceServerFuncWithConfig(tags map[string]string, config TraceServerFuncConfig) echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			if config.Skipper(c) {
+				return next(c)
+			}
+			sc := t.tracer.Extract(b3.ExtractHTTP(c.Request()))
+			span := t.tracer.StartSpan(fmt.Sprintf("S %s %s", c.Request().Method, c.Request().URL.Path), zipkin.Parent(sc))
+			for key, value := range tags {
+				span.Tag(key, value)
+			}
+			defer span.Finish()
+			ctx := zipkin.NewContext(c.Request().Context(), span)
+			c.SetRequest(c.Request().WithContext(ctx))
+			nrw := NewResponseWriter(c.Response().Writer)
+			next(c)
+
+			if nrw.Size() > 0 {
+				zipkin.TagHTTPResponseSize.Set(span, strconv.FormatInt(int64(nrw.Size()), 10))
+			}
+			if nrw.Status() < 200 || nrw.Status() > 299 {
```

Comment:
when `next(c)` would return an error.  `nrw.Status()` is `200`.  This is because how Echo handlers and middleware work -  when you return an error it bubbles up until it discarded or  global error handler is reached and error response is sent.

Reply:
NB: in case of an error `nrw.Size()` == 0 and response is sent to client from global error handler.